app/redux/ducks/classify.js

```diff
@@ -346,6 +363,42 @@ export function saveAnnotations(annotations) {
   };
 }
 
+export function loadWorkflow(workflowId, locale, preferences) {
+  return (dispatch) => {
+    if (preferences) {
+      preferences.update({ 'preferences.selected_workflow': workflowId });
+    }
+    dispatch(reset());
+    dispatch({
+      type: FETCH_WORKFLOW,
+      payload: {
+        workflowId,
+        locale
+      }
+    });
+    const awaitTranslation = dispatch(translations.load('workflow', workflowId, locale));
+    return Promise.all([awaitWorkflow(workflowId), awaitTranslation])
+    .then(([workflow]) => {
```

Comment:
Is that something that could happen when this action is called from  the workflow assignment dialog? I’m assuming you can’t try to load an inactive workflow from the project home.

Reply:
Looking at the API call, it should error with a 404 status if there’s no workflow, but I haven’t tested that to make sure.

Reply:
Yeah, it's possible from workflow assignment since whether or not those ids in the UPP are active isn't checked until we do the request (this eventually could be refactored to avoid the API request). I thought the activeness param would return an empty array and 200 if the workflow requested is no longer active, though I could be misremembering, and this might be the behavior for workflows you do not have permission to (workflows in development on a non-live project?). Regardless, I think there should be a existence check here before saving UPP and dispatching to the Redux store.

Reply:
I looked it over and lines 115-136 in the workflow selection component are what checks for the existence and has slightly different handling than the catch. I think I was correct in thinking that these are inactive workflows being requested for. Could be from ids stored in UPP, and possibly other cases. There were cases of inactive workflows still being project defaults before the lab got fixed too.

Reply:
With the current code, if you are promoted to an inactive workflow, the classifier will simply load the workflow by ID without giving any indication that it's inactive. However, you won't be able to select the new workflow from the project home page.

Reply:
AH, I see what you mean. I think that checking against the serialized active workflow ids array was added a bit later for all of the workflow ids. 

I believe I added the existence safety check prior to the serialization of the active workflow links. What I'm suggesting, though, is that we still do an existence check in the response and add handling if the response array is empty because it is possible as I've run into situations that I didn't think of and workflow selection breaking because of it. I can't see that happening reading it over again, but I've been burned in the past couple of years from some buggy situations.

Reply:
Ok, but i think that an empty response from `GET /workflows/:id` would be a bug in Panoptes. Requesting a specific resource should return something or an error.

Reply:
It occurred to me that `GET /workflows/?workflow_id={:id}&active=true` would return an empty array for inactive workflows, since searches return arrays, but this code doesn't use that route. Neither does workflow selection. We removed that from workflow selection because the code should request a workflow directly when the ID is known, rather than running a search across the workflows table.

This code and workflow selection use an endpoint that returns either a single resource or a 404 error, and the 404 is caught by the catch block.

There's also a bug in workflow selection where API errors run the error handling code twice. I'll open an issue for that.

Reply:
I've updated the `then` block to raise a 404 error, and clear stored preferences, if the workflow doesn't exist. If that case does happen, we really should look at why it's happening and fix Panoptes.

Reply:
Good point on the search param version of the request. Looking back through commit history, that's what we used to use, so it made sense then to check whether or not the response array is empty.

Reply:
From memory, the original code caused performance problems when Panoptes was under heavy load. 

It’s returning an array now because `Promise.all` returns `[workflow, translation]`. I debated adding the translation to make the code easier to understand, but left it out because the linter complains about declaring variables that aren’t used.